projects/storefrontlib/src/cms-components/cart/cart-coupon/cart-coupon.component.spec.ts

```diff
@@ -0,0 +1,125 @@
+    'getActive',
+    'addVoucher',
+    'getAddVoucherResultSuccess',
+    'resetAddVoucherProcessingState',
+    'getAddVoucherResultLoading',
+    'getLoaded',
+  ]);
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      imports: [I18nTestingModule, ReactiveFormsModule],
+      declarations: [
+        CartCouponComponent,
+        MockAppliedCouponsComponent,
+        MockCxIconComponent,
+      ],
+      providers: [
+        { provide: CartService, useValue: mockCartService },
+        CartCouponAnchorService,
+      ],
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(CartCouponComponent);
+    component = fixture.componentInstance;
+    el = fixture.debugElement;
+    cartCouponAnchorService = TestBed.get(CartCouponAnchorService);
+    spyOn(cartCouponAnchorService, 'getEventEmit').and.returnValue(emitter);
+
+    mockCartService.getActive.and.returnValue(of<Cart>({ code: '123' }));
+    mockCartService.getLoaded.and.returnValue(of(true));
+    mockCartService.getAddVoucherResultSuccess.and.returnValue(of());
+    mockCartService.getAddVoucherResultLoading.and.returnValue(of());
+    mockCartService.addVoucher.and.stub();
+    mockCartService.resetAddVoucherProcessingState.and.stub();
+    mockCartService.resetAddVoucherProcessingState.calls.reset();
+  });
+
+  it('should create', () => {
+    fixture.detectChanges();
+    expect(component).toBeTruthy();
+  });
+
+  it('should show coupon input and submit buttom', () => {
+    fixture.detectChanges();
+    expect(el.query(By.css('[data-test="title-coupon"]'))).toBeTruthy();
+    expect(el.query(By.css('[data-test="input-coupon"]'))).toBeTruthy();
+    expect(el.query(By.css('[data-test="button-coupon"]'))).toBeTruthy();
+    expect(
+      el.query(By.css('[data-test="button-coupon"]')).nativeElement.disabled
+    ).toBeTruthy();
+  });
+
+  it('should disable button when cart is loading', () => {
+    mockCartService.getAddVoucherResultLoading.and.returnValue(of(true));
+    fixture.detectChanges();
+
+    input = el.query(By.css('[data-test="input-coupon"]')).nativeElement;
+    input.value = 'couponCode1';
+    input.dispatchEvent(new Event('input'));
+    el.query(By.css('[data-test="button-coupon"]')).nativeElement.click();
+
+    expect(
+      el.query(By.css('[data-test="button-coupon"]')).nativeElement.disabled
+    ).toBeTruthy();
+  });
+
+  it('should apply coupon', () => {
+    mockCartService.getAddVoucherResultLoading.and.returnValue(of(true));
+    mockCartService.getAddVoucherResultSuccess.and.returnValue(of(true));
+
+    fixture.detectChanges();
+
+    input = el.query(By.css('[data-test="input-coupon"]')).nativeElement;
+    input.value = 'couponCode1';
+    input.dispatchEvent(new Event('input'));
+    el.query(By.css('[data-test="button-coupon"]')).nativeElement.click();
+
+    expect(
+      el.query(By.css('[data-test="button-coupon"]')).nativeElement.disabled
+    ).toBeTruthy();
+  });
+});
```

Comment:
add 1 more case:
1. if apply coupon failed, what's the status of textbox and button?

Reply:
and as we aligned, should call ondestroy manually if it existed

Reply:
> 2. click anchor, then the textbox should be focus

As we discussed before, these 2 components communicate by event, so the anchor cannot be get in coupon component. Regarding this case, it is very difficult to test.

Reply:
Done